modules/javafx.graphics/src/main/java/javafx/scene/layout/Region.java

```diff
@@ -2058,62 +2028,133 @@ public final double snappedRightInset() {
         }
     }
 
+    /*
+     * Definition of used terms:
+     *
+     * # available width/heights:
+     *
+     * Sizes provided by the container that may be used as a dependent value when
+     * calculating sizes for biased controls. These may be set to -1 to indicate
+     * no such information is available. If given, the sizes include the Margin
+     * of the child. As such the Margin must be removed before passing these
+     * values as a dependent value to min/pref/max width/height functions.
+     *
+     * # content width/heights:
+     *
+     * The space allocated to a child, minus its margins. These are never -1.
+     *
+     * # bounded width/heights:
+     *
+     * The space allocated to a child, minus its margins, adjusted according to
+     * its constraints (min <= X <= max). These are never -1.
+     */
+
+    /*
+     * Given a content width, limits it by the child's constraints. The fill boolean
+     * controls whether the content width or the child's preferred width is used to compute
+     * the bounded width.
+     */
+    private double computeBoundedWidth(Node child, boolean fill, double contentWidth) {
+        double min = child.minWidth(-1);
+        double max = child.maxWidth(-1);
+
+        if (fill) {
+            return snapSizeX(boundedSize(min, contentWidth, max));
+        }
+
+        return snapSizeX(boundedSize(min, child.prefWidth(-1), Math.min(max, contentWidth)));
+    }
+
+    /*
+     * Given a content height, limits it by the child's constraints. The fill boolean
+     * controls whether the content height or the child's preferred height is used to compute
+     * the bounded height.
+     */
+    private double computedBoundedHeight(Node child, boolean fill, double contentHeight) {
+        double min = child.minHeight(-1);
+        double max = child.maxHeight(-1);
+
+        if (fill) {
+            return snapSizeY(boundedSize(min, contentHeight, max));
+        }
+
+        return snapSizeY(boundedSize(min, child.prefHeight(-1), Math.min(max, contentHeight)));
+    }
+
+    /*
+     * Removes the given Margin (if any) from a width which still includes margins
+     * to create a content width.
+     */
+    private double computeContentWidth(Insets margin, double width) {
+        boolean snap = isSnapToPixel();
+        double left = margin != null ? snapSpaceX(margin.getLeft(), snap) : 0;
+        double right = margin != null ? snapSpaceX(margin.getRight(), snap) : 0;
+
+        return width - left - right;
```

Comment:
snap?

Reply:
Yes, sorry, this was addressed in the larger discussion.  I've left this as-is to keep the PR focus'd on one thing.

The calculation here is using 3 snapped values, and one can reasonably assume the result is "nearly" snapped.  If this value is used later with a ceiling function though, then it might ceil to the next higher value if the result is slightly too high due to floating point errors.  This is why it might be a good idea to adjust how our ceiling functions work in all cases; instead of using a tiny epsilon (or ulp), use a much larger value but still tiny in terms of pixels (like 1/10000th of a pixel).  Any "near" snapped values won't accidentally get rounded up to the next higher pixel then when ceil is used.

Reply:
> Any "near" snapped values won't accidentally get rounded up to the next higher pixel

I like this idea!

Reply:
I agree. This sounds promising.